File: slideshow-go/slideshow.go
package main

import (
	"fmt"
	"hash/fnv"
	"image"
	"image/color"
	"image/draw"
	"io"
	"math/rand"
	"time"
)

type desktop interface {
	Size() image.Point
	ScreenSize() image.Point
	Present(*image.NRGBA) error
	Fullscreen(bool)
	Poll() ([]string, error)
	Close()
}

type shuffleBag struct {
	files, pending []string
	last           string
	random         *rand.Rand
}

func newShuffleBag(files []string, seed *string) *shuffleBag {
	if len(files) == 0 {
		panic("A shuffle bag needs at least one image")
	}
	s := time.Now().UnixNano()
	if seed != nil {
		h := fnv.New64a()
		_, _ = h.Write([]byte(*seed))
		s = int64(h.Sum64())
	}
	return &shuffleBag{files: append([]string(nil), files...), random: rand.New(rand.NewSource(s))}
}
func (b *shuffleBag) next() string {
	if len(b.pending) == 0 {
		b.pending = append([]string(nil), b.files...)
		b.random.Shuffle(len(b.pending), func(i, j int) { b.pending[i], b.pending[j] = b.pending[j], b.pending[i] })
		n := len(b.pending) - 1
		if n > 0 && b.pending[n] == b.last {
			b.pending[0], b.pending[n] = b.pending[n], b.pending[0]
		}
	}
	n := len(b.pending) - 1
	b.last = b.pending[n]
	b.pending = b.pending[:n]
	return b.last
}

type slideshow struct {
	w                          desktop
	bag                        *shuffleBag
	interval                   time.Duration
	history                    []string
	position                   int
	bad                        map[string]bool
	current                    string
	photo                      *image.NRGBA
	background                 color.NRGBA
	paused, fullscreen, closed bool
	exitCode                   int
	deadline, resizeDeadline   time.Time
	stderr                     io.Writer
	clock                      func() time.Time
}

func newSlideshow(w desktop, files []string, o options, bg color.NRGBA, stderr io.Writer) *slideshow {
	a := &slideshow{w: w, bag: newShuffleBag(files, o.seed), interval: o.interval, position: -1, bad: map[string]bool{}, background: bg, fullscreen: !o.windowed, stderr: stderr, clock: time.Now}
	w.Fullscreen(a.fullscreen)
	return a
}
func (a *slideshow) restartTimer(now time.Time) {
	a.deadline = time.Time{}
	if !a.paused && !a.closed {
		a.deadline = now.Add(a.interval)
	}
}
func (a *slideshow) display(path string) bool {
	size, screen := a.w.Size(), a.w.ScreenSize()
	size.X, size.Y = max(1, size.X), max(1, size.Y)
	photo, err := loadImage(path, image.Pt(max(1, min(size.X, screen.X)), max(1, min(size.Y, screen.Y))))
	if err != nil {
		fmt.Fprintf(a.stderr, "Skipping %s: %v\n", path, err)
		a.bad[path] = true
		return false
	}
	frame := image.NewNRGBA(image.Rect(0, 0, size.X, size.Y))
	draw.Draw(frame, frame.Bounds(), image.NewUniform(a.background), image.Point{}, draw.Src)
	offset := image.Pt((size.X-photo.Bounds().Dx())/2, (size.Y-photo.Bounds().Dy())/2)
	draw.Draw(frame, photo.Bounds().Add(offset), photo, image.Point{}, draw.Over)
	if err := a.w.Present(frame); err != nil {
		fmt.Fprintln(a.stderr, "Desktop error:", err)
		a.exitCode = 1
		a.close()
		return false
	}
	a.photo, a.current = photo, path
	return true
}
func (a *slideshow) nextImage(now time.Time) {
	if a.closed {
		return
	}
	for a.position+1 < len(a.history) {
		a.position++
		path := a.history[a.position]
		if !a.bad[path] && a.display(path) {
			a.restartAfterDisplay(now)
			return
		}
		if a.closed {
			return
		}
	}
	for i := 0; i < 2*len(a.bag.files); i++ {
		path := a.bag.next()
		if !a.bad[path] && a.display(path) {
			a.history = append(a.history, path)
			a.position = len(a.history) - 1
			a.restartAfterDisplay(now)
			return
		}
		if a.closed {
			return
		}
	}
	fmt.Fprintln(a.stderr, "No readable images remain; exiting.")
	a.exitCode = 1
	a.close()
}

// Start after decoding so slow files still get a full interval on screen.
func (a *slideshow) restartAfterDisplay(now time.Time) { a.restartTimer(a.clock()) }
func (a *slideshow) previousImage(now time.Time) {
	original := a.position
	for a.position > 0 {
		a.position--
		path := a.history[a.position]
		if !a.bad[path] && a.display(path) {
			a.restartAfterDisplay(now)
			return
		}
		if a.closed {
			return
		}
	}
	a.position = original
	a.restartTimer(now)
}
func (a *slideshow) handle(key string, now time.Time) {
	if a.closed {
		return
	}
	switch key {
	case "Escape", "q", "Q", "close":
		a.close()
	case "space":
		a.paused = !a.paused
		a.restartTimer(now)
	case "Right", "n", "N", "click":
		a.nextImage(now)
	case "Left", "p", "P":
		a.previousImage(now)
	case "f", "F", "F11":
		a.fullscreen = !a.fullscreen
		a.w.Fullscreen(a.fullscreen)
		a.resizeDeadline = now.Add(100 * time.Millisecond)
	case "resize":
		a.resizeDeadline = now.Add(100 * time.Millisecond)
	case "expose":
		if a.current != "" && !a.display(a.current) {
			a.nextImage(now)
		}
	}
}
func (a *slideshow) tick(now time.Time) {
	if a.closed {
		return
	}
	if !a.resizeDeadline.IsZero() && !now.Before(a.resizeDeadline) {
		a.resizeDeadline = time.Time{}
		if a.current != "" && !a.display(a.current) {
			a.nextImage(now)
		}
	}
	if !a.deadline.IsZero() && !now.Before(a.deadline) {
		a.nextImage(now)
	}
}
func (a *slideshow) close() {
	a.closed = true
	a.deadline = time.Time{}
	a.resizeDeadline = time.Time{}
}

File: slideshow/slideshow.py
"""A full-screen, shuffled photo slideshow. Run with --help for options."""

import argparse
import math
from pathlib import Path
import random
import sys
import tkinter as tk

from PIL import Image, ImageOps, ImageTk


EXTENSIONS = {".jpg", ".jpeg", ".png", ".webp", ".gif", ".bmp", ".tif", ".tiff"}


def positive_seconds(value):
    try:
        seconds = float(value)
    except ValueError as exc:
        raise argparse.ArgumentTypeError("interval must be a positive number") from exc
    if not math.isfinite(seconds) or seconds <= 0 or seconds > 2147483:
        raise argparse.ArgumentTypeError("interval must be finite and between 0 (exclusive) and 2147483 seconds")
    return seconds


def discover(directory, recursive=False):
    directory = Path(directory).expanduser().resolve()
    if not directory.is_dir():
        raise ValueError(f"Not a directory: {directory}")
    files = sorted(path for path in (directory.rglob("*") if recursive else directory.iterdir())
                   if path.is_file() and path.suffix.lower() in EXTENSIONS)
    if not files:
        raise ValueError(f"No supported images found in: {directory}")
    return files


class ShuffleBag:
    def __init__(self, files, seed=None):
        self.files = list(files)
        if not self.files:
            raise ValueError("A shuffle bag needs at least one image")
        self.random = random.Random(seed)
        self.pending = []
        self.last = None

    def __next__(self):
        if not self.pending:
            self.pending = self.files.copy()
            self.random.shuffle(self.pending)
            if len(self.pending) > 1 and self.pending[-1] == self.last:
                self.pending[0], self.pending[-1] = self.pending[-1], self.pending[0]
        self.last = self.pending.pop()
        return self.last


def load_image(path, size):
    with Image.open(path) as source:
        oriented = ImageOps.exif_transpose(source)
        # Animated/multipage files display their first frame only.
        return ImageOps.contain(oriented.convert("RGBA"), size, Image.Resampling.LANCZOS)


class Slideshow:
    def __init__(self, root, files, interval=5, background="black", seed=None, windowed=False):
        self.root = root
        self.bag = ShuffleBag(files, seed)
        self.interval_ms = max(1, round(interval * 1000))
        self.history = []
        self.position = -1
        self.bad_files = set()
        self.photo = None
        self.current_path = None
        self.paused = False
        self.fullscreen = not windowed
        self.timer = None
        self.resize_timer = None
        self.closed = False
        self.exit_code = 0
        root.title("Photo slideshow")
        root.geometry("1280x720")
        self.canvas = tk.Canvas(root, background=background, highlightthickness=0)
        self.canvas.pack(fill="both", expand=True)
        self.image_item = self.canvas.create_image(0, 0, anchor="center")
        self.apply_fullscreen()
        for key in ("<Escape>", "q", "Q"):
            root.bind(key, self.close)
        root.bind("<space>", self.toggle_pause)
        for key in ("<Right>", "n", "N", "<Button-1>"):
            root.bind(key, self.next_image)
        for key in ("<Left>", "p", "P"):
            root.bind(key, self.previous_image)
        for key in ("f", "F", "<F11>"):
            root.bind(key, self.toggle_fullscreen)
        self.canvas.bind("<Configure>", self.on_resize)
        root.protocol("WM_DELETE_WINDOW", self.close)
        root.update_idletasks()
        self.next_image()

    def apply_fullscreen(self):
        self.root.attributes("-fullscreen", self.fullscreen)
        self.root.configure(cursor="none" if self.fullscreen else "")

    def toggle_fullscreen(self, event=None):
        self.fullscreen = not self.fullscreen
        self.apply_fullscreen()

    def restart_timer(self):
        if self.timer is not None:
            self.root.after_cancel(self.timer)
            self.timer = None
        if not self.paused and not self.closed:
            self.timer = self.root.after(self.interval_ms, self.next_image)

    def toggle_pause(self, event=None):
        self.paused = not self.paused
        self.restart_timer()

    def display(self, path):
        # Query dimensions on every display; canvas dimensions also follow resizing.
        screen = (self.root.winfo_screenwidth(), self.root.winfo_screenheight())
        width = max(1, self.canvas.winfo_width())
        height = max(1, self.canvas.winfo_height())
        size = (min(width, screen[0]), min(height, screen[1]))
        try:
            resized = load_image(path, size)
        except (OSError, ValueError, SyntaxError, Image.DecompressionBombError) as exc:
            print(f"Skipping {path}: {exc}", file=sys.stderr)
            self.bad_files.add(path)
            return False
        self.photo = ImageTk.PhotoImage(resized, master=self.root)
        self.canvas.itemconfigure(self.image_item, image=self.photo)
        self.canvas.coords(self.image_item, width / 2, height / 2)
        self.current_path = path
        return True

    def next_image(self, event=None):
        while self.position + 1 < len(self.history):
            self.position += 1
            path = self.history[self.position]
            if path not in self.bad_files and self.display(path):
                self.restart_timer()
                return
        # Two bag lengths cover a partial cycle and a full cycle, even after failures.
        for _ in range(2 * len(self.bag.files)):
            path = next(self.bag)
            if path not in self.bad_files and self.display(path):
                self.history.append(path)
                self.position = len(self.history) - 1
                self.restart_timer()
                return
        print("No readable images remain; exiting.", file=sys.stderr)
        self.exit_code = 1
        self.close()

    def previous_image(self, event=None):
        original = self.position
        while self.position > 0:
            self.position -= 1
            path = self.history[self.position]
            if path not in self.bad_files and self.display(path):
                self.restart_timer()
                return
        self.position = original
        self.restart_timer()

    def on_resize(self, event=None):
        if self.resize_timer is not None:
            self.root.after_cancel(self.resize_timer)
        self.resize_timer = self.root.after(100, self.redraw)

    def redraw(self):
        self.resize_timer = None
        if self.current_path is not None and not self.display(self.current_path):
            self.next_image()

    def close(self, event=None):
        self.closed = True
        for timer in (self.timer, self.resize_timer):
            if timer is not None:
                self.root.after_cancel(timer)
        self.timer = self.resize_timer = None
        self.root.destroy()


def main(argv=None):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("directory", help="Directory containing images")
    parser.add_argument("-i", "--interval", type=positive_seconds, default=5, metavar="SECONDS")
    parser.add_argument("--background", default="black", metavar="COLOR")
    parser.add_argument("--recursive", action="store_true")
    parser.add_argument("--seed")
    parser.add_argument("--windowed", action="store_true")
    args = parser.parse_args(argv)
    try:
        files = discover(args.directory, args.recursive)
    except (ValueError, OSError) as exc:
        parser.error(str(exc))
    root = None
    try:
        root = tk.Tk()
        root.winfo_rgb(args.background)
        app = Slideshow(root, files, args.interval, args.background, args.seed, args.windowed)
        if not app.closed:
            root.mainloop()
        return app.exit_code
    except tk.TclError as exc:
        print(f"Cannot open slideshow: {exc}. Check the desktop/DISPLAY and background color.", file=sys.stderr)
        if root is not None:
            try:
                root.destroy()
            except tk.TclError:
                pass
        return 1


if __name__ == "__main__":
    sys.exit(main())
